In [ ]:
from tensorflow.keras.datasets import imdb
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=10000)
(X_train, y_train), (X_test, y_test) 

2025-09-04 15:06:19.214770: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2025-09-04 15:06:20.099908: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


17464789/17464789 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step


((array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
         list([1, 194, 1153, 194, 8255, 78, 

In [7]:
# 텍스트 데이터의 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad = pad_sequences(X_train, maxlen=80, padding='post', truncating='post')
X_test_pad = pad_sequences(X_test, maxlen=80, padding='post', truncating='post')

In [4]:
from tensorflow.keras import Sequential, layers
model_dnn = Sequential([
    layers.Input(shape=(80,)), # 80개의 단어를 입력 받는 것
    layers.Embedding(input_dim=10000, output_dim=32), # (80, 32)
    layers.Flatten(),
    layers.Dense(64, activation='relu'),
    layers.Dense(2, activation='softmax')
])

2025-09-04 15:14:39.202350: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:984] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node
Your kernel may have been built without NUMA support.
2025-09-04 15:14:39.392693: W tensorflow/core/common_runtime/gpu/gpu_device.cc:2251] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...


In [5]:
model_dnn.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_1 (Embedding)         │ (None, 80, 32)         │       320,000 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_1 (Flatten)             │ (None, 2560)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 64)             │       163,904 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 2)              │           130 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 484,034 (1.85 MB)

 Trainable params: 484,034 (1.85 MB)

 Non-trainable params: 0 (0.00 B)

In [9]:
model_dnn.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_dnn.fit(X_train_pad, y_train, epochs=10, batch_size=200)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.6788 - loss: 0.5765
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8844 - loss: 0.2851
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9795 - loss: 0.0839
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9987 - loss: 0.0143
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9999 - loss: 0.0036
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 1.0000 - loss: 0.0017
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 1.0000 - loss: 0.0011
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 1.0000 - loss: 7.7184e-04
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 1.0000 - loss: 5.7468e-04
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 1.0000 - loss: 4.4550e-04


In [10]:
model_dnn.evaluate(X_test_pad, y_test)

782/782 ━━━━━━━━━━━━━━━━━━━━ 1s 984us/step - accuracy: 0.7783 - loss: 0.9000


[0.9000036716461182, 0.7783200144767761]

In [15]:
# RNN 모델
from tensorflow.keras import Sequential, layers
model_rnn = Sequential([
    layers.Input(shape=(80,)), # 80개의 단어를 입력 받는 것
    layers.Embedding(input_dim=10000, output_dim=32), # (80, 32)
    # layers.Flatten(),
    # layers.Dense(64, activation='relu'),
    layers.SimpleRNN(64, return_sequences=True),
    layers.SimpleRNN(128),
    layers.Dense(2, activation='softmax')
])

In [16]:
model_rnn.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_4 (Embedding)         │ (None, 80, 32)         │       320,000 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ simple_rnn_2 (SimpleRNN)        │ (None, 80, 64)         │         6,208 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ simple_rnn_3 (SimpleRNN)        │ (None, 128)            │        24,704 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 2)              │           258 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 351,170 (1.34 MB)

 Trainable params: 351,170 (1.34 MB)

 Non-trainable params: 0 (0.00 B)

In [17]:
model_rnn.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 6s 42ms/step - accuracy: 0.6512 - loss: 0.6025
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 42ms/step - accuracy: 0.8367 - loss: 0.3831
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 42ms/step - accuracy: 0.8872 - loss: 0.2842
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.9356 - loss: 0.1775
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.9723 - loss: 0.0832
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 43ms/step - accuracy: 0.9884 - loss: 0.0371
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 41ms/step - accuracy: 0.9901 - loss: 0.0278
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.9944 - loss: 0.0176
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.9945 - loss: 0.0171
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 43ms/step - accuracy: 0.9952 - loss: 0.0137
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.7355 - loss: 1.4589


[1.4588607549667358, 0.7355200052261353]

전\
[0.9000036716461182, 0.7783200144767761]

후\
[1.4588607549667358, 0.7355200052261353]

[loss, accuracy] => 둘 다 안좋아짐

안 좋아진 이유: 경사소실, 과대적합, ...

# RNN으로 영화평 분류하기

## 1. 옵티마이저를 sgd로 바꿔보세요.

In [ ]:
from tensorflow.keras import Sequential, layers
model_rnn2 = Sequential([
    layers.Input(shape=(80,)), # 80개의 단어를 입력 받는 것
    layers.Embedding(input_dim=10000, output_dim=32), # (80, 32)
    # layers.Flatten(),
    # layers.Dense(64, activation='relu'),
    layers.SimpleRNN(64, return_sequences=True),
    layers.SimpleRNN(128),
    layers.Dense(2, activation='softmax')
])

In [ ]:
model_rnn2.compile(optimizer='sgd', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn2.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn2.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 6s 40ms/step - accuracy: 0.4950 - loss: 0.6994
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.4973 - loss: 0.6949
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.5032 - loss: 0.6945
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.5072 - loss: 0.6937
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.5051 - loss: 0.6941
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.5080 - loss: 0.6938
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.5124 - loss: 0.6936
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.5084 - loss: 0.6935
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.5124 - loss: 0.6932
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.5133 - loss: 0.6929
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.5002 - loss: 0.6954


[0.6954269409179688, 0.5002400279045105]

전\
[0.9000036716461182, 0.7783200144767761]

후\
[0.6954269409179688, 0.5002400279045105]

loss -> 좋아짐, accuracy -> 안좋아짐

## 2. 전체 단어의 개수를 1000개로 바꿔보세요.

In [42]:
from tensorflow.keras.datasets import imdb
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=1000)
(X_train, y_train), (X_test, y_test)

((array([list([1, 14, 22, 16, 43, 530, 973, 2, 2, 65, 458, 2, 66, 2, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 2, 2, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2, 19, 14, 22, 4, 2, 2, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 2, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2, 2, 16, 480, 66, 2, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 2, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 2, 15, 256, 4, 2, 7, 2, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 2, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2, 56, 26, 141, 6, 194, 2, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 2, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 2, 88, 12, 16, 283, 5, 16, 2, 113, 103, 32, 15, 16, 2, 19, 178, 32]),
         list([1, 194, 2, 194, 2, 78, 228, 5, 6, 2, 2, 2, 134, 26, 4, 715, 8, 118, 2, 14, 394, 20, 13, 119, 954, 

In [43]:
# 텍스트 데이터의 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad = pad_sequences(X_train, maxlen=80, padding='post', truncating='post')
X_test_pad = pad_sequences(X_test, maxlen=80, padding='post', truncating='post')

In [44]:
from tensorflow.keras import Sequential, layers
model_rnn3 = Sequential([
    layers.Input(shape=(80,)), # 80개의 단어를 입력 받는 것
    layers.Embedding(input_dim=1000, output_dim=32), # (80, 32)
    # layers.Flatten(),
    # layers.Dense(64, activation='relu'),
    layers.SimpleRNN(64, return_sequences=True),
    layers.SimpleRNN(128),
    layers.Dense(2, activation='softmax')
])

In [45]:
model_rnn3.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn3.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn3.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 7s 41ms/step - accuracy: 0.5094 - loss: 0.6986
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.5871 - loss: 0.6660
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.7264 - loss: 0.5372
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.7999 - loss: 0.4378
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.8324 - loss: 0.3840
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.8573 - loss: 0.3368
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.8685 - loss: 0.3097
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.8943 - loss: 0.2592
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.9110 - loss: 0.2238
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.9305 - loss: 0.1800
782/782 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.7226 - loss: 0.9093


[0.9092904925346375, 0.722599983215332]

전\
[0.9000036716461182, 0.7783200144767761]

후\
[0.9092904925346375, 0.722599983215332]

loss -> 안좋아짐, accuracy -> 안좋아짐

## 3. 영화평의 길이를 200개로 바꿔보세요.

In [46]:
from tensorflow.keras.datasets import imdb
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=10000)
(X_train, y_train), (X_test, y_test)

((array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
         list([1, 194, 1153, 194, 8255, 78, 

In [47]:
# 텍스트 데이터의 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad = pad_sequences(X_train, maxlen=80, padding='post', truncating='post')
X_test_pad = pad_sequences(X_test, maxlen=80, padding='post', truncating='post')

In [48]:
from tensorflow.keras import Sequential, layers
model_rnn4 = Sequential([
    layers.Input(shape=(200,)), # 80개의 단어를 입력 받는 것
    layers.Embedding(input_dim=10000, output_dim=32), # (80, 32)
    # layers.Flatten(),
    # layers.Dense(64, activation='relu'),
    layers.SimpleRNN(64, return_sequences=True),
    layers.SimpleRNN(128),
    layers.Dense(2, activation='softmax')
])

In [49]:
model_rnn4.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn4.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn4.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 6s 41ms/step - accuracy: 0.5480 - loss: 0.6812
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.8062 - loss: 0.4363
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step - accuracy: 0.9024 - loss: 0.2491
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.9698 - loss: 0.0882
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.9853 - loss: 0.0458
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 41ms/step - accuracy: 0.9891 - loss: 0.0318
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 41ms/step - accuracy: 0.9913 - loss: 0.0274
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.9911 - loss: 0.0260
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.9966 - loss: 0.0100
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 41ms/step - accuracy: 0.9928 - loss: 0.0231
782/782 ━━━━━━━━━━━━━━━━━━━━ 4s 6ms/step - accuracy: 0.7470 - loss: 1.3514


[1.3514418601989746, 0.746999979019165]

전\
[0.9000036716461182, 0.7783200144767761]

후\
[1.3514418601989746, 0.746999979019165]

loss -> 안좋아짐, accuracy -> 안좋아짐

## 4. pad_sequence의 truncating과 padding을 pre로 바꿔보세요.

In [58]:
from tensorflow.keras.datasets import imdb
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=10000)
(X_train, y_train), (X_test, y_test)

((array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
         list([1, 194, 1153, 194, 8255, 78, 

In [59]:
# 텍스트 데이터의 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad = pad_sequences(X_train, maxlen=80, padding='pre', truncating='pre')
X_test_pad = pad_sequences(X_test, maxlen=80, padding='pre', truncating='pre')

In [60]:
from tensorflow.keras import Sequential, layers
model_rnn5 = Sequential([
    layers.Input(shape=(80,)), # 80개의 단어를 입력 받는 것
    layers.Embedding(input_dim=10000, output_dim=32), # (80, 32)
    # layers.Flatten(),
    # layers.Dense(64, activation='relu'),
    layers.SimpleRNN(64, return_sequences=True),
    layers.SimpleRNN(128),
    layers.Dense(2, activation='softmax')
])

In [61]:
model_rnn5.compile(optimizer='sgd', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn5.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn5.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 6s 41ms/step - accuracy: 0.5011 - loss: 0.7010
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 41ms/step - accuracy: 0.5073 - loss: 0.6940
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.5116 - loss: 0.6936
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.5210 - loss: 0.6928
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 42ms/step - accuracy: 0.5202 - loss: 0.6918
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 42ms/step - accuracy: 0.5263 - loss: 0.6911
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 42ms/step - accuracy: 0.5305 - loss: 0.6900
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 42ms/step - accuracy: 0.5422 - loss: 0.6880
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 41ms/step - accuracy: 0.5472 - loss: 0.6860
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 40ms/step - accuracy: 0.5570 - loss: 0.6841
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.5113 - loss: 0.6988


[0.6988484859466553, 0.5113199949264526]

전\
[0.9000036716461182, 0.7783200144767761]

후\
[0.6845731735229492, 0.5113199949264526]

loss -> 좋아짐, accuracy -> 안좋아짐

## 5. RNN 층(뉴런 128개)을 하나 더 추가해 보세요.

In [54]:
from tensorflow.keras.datasets import imdb
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=10000)
(X_train, y_train), (X_test, y_test)

((array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
         list([1, 194, 1153, 194, 8255, 78, 

In [55]:
# 텍스트 데이터의 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad = pad_sequences(X_train, maxlen=80, padding='post', truncating='post')
X_test_pad = pad_sequences(X_test, maxlen=80, padding='post', truncating='post')

In [56]:
from tensorflow.keras import Sequential, layers
model_rnn6 = Sequential([
    layers.Input(shape=(80,)), # 80개의 단어를 입력 받는 것
    layers.Embedding(input_dim=10000, output_dim=32), # (80, 32)
    # layers.Flatten(),
    # layers.Dense(64, activation='relu'),
    layers.SimpleRNN(64, return_sequences=True),
    layers.SimpleRNN(128, return_sequences=True),
    layers.SimpleRNN(128),
    layers.Dense(2, activation='softmax')
])

In [57]:
model_rnn6.compile(optimizer='sgd', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn6.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn6.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 9s 69ms/step - accuracy: 0.5086 - loss: 0.7046
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 9s 70ms/step - accuracy: 0.5145 - loss: 0.6943
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 9s 72ms/step - accuracy: 0.5130 - loss: 0.6944
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 9s 71ms/step - accuracy: 0.5195 - loss: 0.6934
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 9s 75ms/step - accuracy: 0.5223 - loss: 0.6923
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 9s 74ms/step - accuracy: 0.5287 - loss: 0.6914
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 9s 75ms/step - accuracy: 0.5272 - loss: 0.6911
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 9s 75ms/step - accuracy: 0.5287 - loss: 0.6904
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 9s 74ms/step - accuracy: 0.5359 - loss: 0.6898
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 9s 74ms/step - accuracy: 0.5343 - loss: 0.6893
782/782 ━━━━━━━━━━━━━━━━━━━━ 7s 9ms/step - accuracy: 0.5187 - loss: 0.6932


[0.6932304501533508, 0.5186799764633179]

전\
[0.9000036716461182, 0.7783200144767761]

후\
[0.6932304501533508, 0.5186799764633179]

loss -> 좋아짐, accuracy -> 안좋아짐

In [ ]:
# RNN 구조의 한계
# 하이퍼파라미터의 튜닝 & EDA

# CNN으로 구성

In [ ]:
model_cnn = Sequential(
    [

        layers.Conv1D(64, 3, activation='relu'),
        layers.GlobalMaxPool1D(),
    ]
)

AttributeError: module 'keras._tf_keras.keras.layers' has no attribute 'SimpleCNN'